R/smooth_mortality_law.R

```diff
@@ -0,0 +1,90 @@
+#' Function to smooth mortality rates using using MortalityLaw package
+#'
+#' This smoothing function allows smoothing of a variable in a vital object using MortalityLaw package.
+#' The vital object is returned along with some additional columns containing
+#' information about the smoothed variable: usually `.smooth` containing the
+#' smoothed values, and `.smooth_se` containing the corresponding standard errors.
+#'
+#' `smooth_mortality_law()` smoothes mortality rates using specified mortality law and optimisation method.
+#' @param .data A vital object
+#' @param .var name of variable to smooth
+#' @param law name of mortality law. For available mortality laws, users can check the \code{\link{MortalityLaws::availableLaws}}
+#' function to learn about the available options.
+#' @param opt.method optimisation method. The models can be fitted under the maximum likelihood methodology
+#' or by selecting a loss function to be optimised. See the implemented loss function by running the \code{\link{MortalityLaws::xavailableLF}} function.
+#' @return vital with added columns containing smoothed values and their standard errors
+#' @keywords smooth, mortality
+#' @rdname smooth_law_vital
+#' @author Sixian Tang and Rob J Hyndman
+#' @examples
+#' library(MortalityLaws)
+#' norway_mortality |> smooth_mortality_law(Mortality)
+#'
+#' @export
+smooth_mortality_law <- function(.data, .var, law = "gompertz", opt.method = "LF2") {
+  smooth_law_vital(.data, {{ .var }}, smooth_mortality_law_x, law, opt.method)
+}
+
+smooth_mortality_law_x <- function(data, var, age, law, opt.method) {
+  y <- data[[var]]
+  x <- data[[age]]
+  smooth.fit <- MortalityLaws::MortalityLaw(x = x, mx = y, law = law, opt.method = opt.method)
+  # Mean squared error
+  residual_variance <- sum((smooth.fit$residuals)^2, na.rm = TRUE) / (length(smooth.fit$fitted.values) - length(smooth.fit$coefficients))
+  standard_error <- sqrt(residual_variance)
+  out <- tibble(
+    age = x,
+    .smooth = smooth.fit$fitted.values,
+    .smooth_se = .smooth * standard_error
```

Comment:
This looks too large. What you call the standard error is the standard deviation of the residuals. So it will not converge to zero as n increases. I suspect this is the appropriate standard deviation if you were constructing a prediction interval, not a confidence interval.